sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapter.java

```diff
@@ -50,17 +53,61 @@
 @Slf4j
 public class LDeviceAdapter extends SclElementAdapter<IEDAdapter, TLDevice> {
 
+    private static final long INTG_PD_VALUE_FOR_FC_MX = 2000L;
+
     /**
      * Constructor
+     *
      * @param parentAdapter Parent container reference
-     * @param currentElem Current reference
+     * @param currentElem   Current reference
      */
     public LDeviceAdapter(IEDAdapter parentAdapter, TLDevice currentElem) {
         super(parentAdapter, currentElem);
     }
 
+    /**
+     * Create DataSet and ReportControl Blocks for the HMI with the given FCDAs.
+     * DataSet and ReportControl are created in LN0, even if FCDA refers to another LN.
+     * @param fcdas List of FCDA for which we must create the DataSet and ReportControl
+     */
+    public void createHmiReportControlBlocks(List<FcdaRecord> fcdas) {
+        LN0Adapter ln0 = getLN0Adapter();
+        if (!ln0.isStatusOn())
+            return;
+        fcdas.stream()
+                .filter(fcda -> getInst().equals(fcda.ldInst()))
+                .forEach(fcda -> {
+                            AbstractLNAdapter<?> sourceLn = TLLN0Enum.LLN_0.value().equals(fcda.lnClass()) ?
+                                    ln0 // ln0 status "ON" has already been checked, no need to check it again
+                                    : findLnAdapter(fcda.lnClass(), fcda.lnInst(), fcda.prefix()).filter(AbstractLNAdapter::isStatusOn).orElse(null);
+                            if (sourceLn != null
+                                    && sourceLn.getDAI(fcda.toResumedDataTemplate(), false)
+                                    .stream().anyMatch(da -> fcda.fc() == da.getFc())) { // getDAI does not filter on DA.fc
+                                createHmiReportCB(ln0, fcda);
+                            }
+                        }
+                );
+    }
+
+    private void createHmiReportCB(LN0Adapter ln0, FcdaRecord fcda) {
+        boolean isFcMx = fcda.fc() == TFCEnum.MX;
+        String dataSetSuffix = getInst().toUpperCase(Locale.ENGLISH) + ATTRIBUTE_VALUE_SEPARATOR + (isFcMx ? "CYPO" : "DQPO");
+        DataSetAdapter dataSet = ln0.createDataSetIfNotExists(DATASET_NAME_PREFIX + dataSetSuffix, ControlBlockEnum.REPORT);
+        dataSet.createFCDAIfNotExists(fcda.ldInst(), fcda.prefix(), fcda.lnClass(), fcda.lnInst(), fcda.doRef(), fcda.daRef(), fcda.fc());
+        String cbName = CONTROLBLOCK_NAME_PREFIX + dataSetSuffix;
+        String cbId = ln0.generateControlBlockId(getLdName(), cbName);
+        String datSet = dataSet.getCurrentElem().getName();
+        ControlBlockAdapter controlBlockAdapter = ln0.createControlBlockIfNotExists(cbName, cbId, datSet, ControlBlockEnum.REPORT);
+        if (isFcMx && controlBlockAdapter.getCurrentElem() instanceof TReportControl tReportControl){
```

Comment:
this check is not needed controlBlockAdapter.getCurrentElem() instanceof TReportControl, you just created it.

Reply:
OK